### PR TITLE
mesonbuild: update to 0.51.2

### PR DESCRIPTION
**Pull request by InuSasha**
since meson 0.51.0, a host build environment ist needed for cross-compiling (instead of target build environment)

build and boot-tested on RPi2

### CODE REVIEW

**Comment by MilhouseVH** [comment]
Wow thanks for this, what a PITA! :smile:

I find the reference to `generic` in `setup_toolchain generic target` a little confusing - I realise it could be anything (other than a valid toolchain ie. meson) but  wonder if we were to flip the args around, maybe we can let the second argument (ie. toolchain) become optional, ie.
```
setup_toolchain target
setup_toolchain host
setup_toolchain target meson
```
What do you think? This would mean the case statement in `setup_toolchain()` becomes:
```
  case ${1}:${2} in
    target:meson)
    target:*)
    host:*)
  esac
```
does that help improve anything, as we no longer need dummy values? I don't know if we will ever need to support additional toolchains (ie. cmake etc.) like this in future (hope not!)

Also, we could add a small proxy helper `create_meson_conf()` since we already pass the target:
```
create_meson_conf()
{
  case ${1} in
    host|bootstrap) create_meson_conf_host ${1} ${2};;
    init|target) create_meson_conf_target ${1} ${2};;
  esac
}
```
then just call `create_meson_conf ${TARGET} ${MESON_CONF}` from scripts/build, it doesn't need to know/care about the host or target differences.

**Comment by InuSasha** [comment]
**setup_toolchain**
the parameter order of `setup_toolchain` change be changed.

**create_meson_conf**
i dropped the proxy, because we do not need this extra step. The proxy will only increase the compexity of code, imho. and we need it, only on one position.

**Comment by MilhouseVH** [comment]
>  the parameter order of setup_toolchain change be changed.
Thanks.

> The proxy will only increase the compexity of code, imho. and we need it, only on one position.
Ok, thanks.

**Comment by InuSasha** [comment]
dropped wrongly changes gtk3

**Comment by MilhouseVH** [comment]
RPi/RPi2/Generic images and all addons built OK with meson-0.51.2. The resulting images all boot and run correctly (as far as I can tell).

I've compared the last published `meson-0.49.1` images (`#0827`) with the new `meson-0.51.2` test images (`#0828`) and there are some fairly significant file size changes (ignore the `pvr.mythtv` noise that crept into the overnight `#0828` test build):

RPi: http://ix.io/1TJe (5.3MB increase with `meson-0.52.1`)
RPi2: http://ix.io/1TJf (5.3MB increase with `meson-0.52.1`)
Generic: http://ix.io/1TJg {9.6MB reduction with `meson-0.52.1`)

For RPi/RPi2 most of the increase comes from `systemd` and `libinput`, and we don't appear to be stripping these `systemd`/`libinput` files when building with `meson-0.51.2`

As for Generic, I can't explain the almost 10MB size reduction... as far as I can tell the Intel i915 driver is working correctly in a Skylake i5 NUC when built with `meson-0.51.2`. Maybe a different `-On` optimisation to what we were using before...?

**Comment by InuSasha** [comment]
look like the striping was not done

**Comment by InuSasha** [comment]
strip is working, now.
ready to push

**Comment by MilhouseVH** [comment]
Thanks @InuSasha - I'll re-run my RPi/RPi2/Generic tests.

File: scripts/build [hunk 1]
```diff
@@ -137,6 +164,10 @@ TARGET_MESON_OPTS="--prefix=/usr \
                    --libexecdir=/usr/lib \
                    --localstatedir=/var \
                    --buildtype=${MESON_BUILD_TYPE}"
+if [ ${BUILD_WITH_DEBUG} != yes ] && flag_enabled "strip" "yes"; then
```
Comment (MilhouseVH) [inline scripts/build]: Tiny nit, just to be consistent with all other instances:
```suggestion
if [ "${BUILD_WITH_DEBUG}" != "yes" ] && flag_enabled "strip" "yes"; then
```

Comment (InuSasha) [reply scripts/build]: Copy'n'paste mistake XD

**Comment by MilhouseVH** [comment]
Looks good now - many thanks!

RPi: http://ix.io/1Uix
RPi2: http://ix.io/1Uiy
Generic: http://ix.io/1Uiz

(the leftmost `#0904` column is for the new `meson-0.51.2` build without GLVND, the rightmost is `meson-0.49.1` with GLVND)

We still get a very nice reduction in the Generic graphic drivers.

I'm going to build all the add-ons for RPi/RPi2/Generic overnight, just to make sure they all build OK.

**Comment by MilhouseVH** [comment]
All add-ons successful!

**Comment by 5schatten** [comment]
@InuSasha could you rebase the mesa package please?

**Comment by InuSasha** [comment]
rebased

**Pull request merged**